File: CP/LeetCode/Python/61.py
# 61. Rotate List

from __future__ import annotations

from typing import Generic, TypeVar

T = TypeVar("T")


class ListNode(Generic[T]):
    def __init__(self, val: T, next: ListNode[T] | None = None):
        self.val = val
        self.next = next
    def __repr__(self) -> str:
        return f"{self.val} -> {self.next}"


class Solution:
    def rotateRight(self, head: ListNode[int], k: int) -> ListNode[int] | None:
        if not head:
            return head
        length = 1
        old_tail = head
        while old_tail.next:
            old_tail = old_tail.next
            length += 1

        old_tail.next = head
        new_tail = head
        for _ in range(length - (k % length) - 1):
            if new_tail.next:
                new_tail = new_tail.next

        new_head = new_tail.next
        new_tail.next = None
        return new_head


if __name__ == "__main__":
    sol = Solution()
    linked_list = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    print(sol.rotateRight(linked_list, 9))

File: CP/LeetCode/C/61.c
// 61. Rotate List

#include <stdio.h>
#include <stdlib.h>

struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *rotateRight(struct ListNode *head, int k)
{
    if (head == NULL || head->next == NULL)
    {
        return head;
    }
    struct ListNode *temp = head;
    int len = 1;
    while (temp->next != NULL)
    {
        temp = temp->next;
        len++;
    }
    temp->next = head;
    k = k % len;
    k = len - k;
    while (k--)
    {
        temp = temp->next;
    }
    head = temp->next;
    temp->next = NULL;
    return head;
}

void free_list(struct ListNode *head)
{
    struct ListNode *temp;
    while (head != NULL)
    {
        temp = head;
        head = head->next;
        free(temp);
        temp = NULL;
    }
}

int main()
{
    struct ListNode *head = (struct ListNode *)malloc(sizeof(struct ListNode));
    head->val = 1;
    head->next = (struct ListNode *)malloc(sizeof(struct ListNode));
    head->next->val = 2;
    head->next->next = (struct ListNode *)malloc(sizeof(struct ListNode));
    head->next->next->val = 3;
    head->next->next->next = (struct ListNode *)malloc(sizeof(struct ListNode));
    head->next->next->next->val = 4;
    head->next->next->next->next = (struct ListNode *)malloc(sizeof(struct ListNode));
    head->next->next->next->next->val = 5;
    head->next->next->next->next->next = NULL;

    int k = 2;
    head = rotateRight(head, k);
    struct ListNode *temp = head;
    while (temp != NULL)
    {
        printf("%d ", temp->val);
        temp = temp->next;
    }
    free(temp);
    free_list(head);
    return 0;
}
